Support time comparisons

serogers opened the pull request:
Really enjoy this gem DRYing up codebases! I've also found myself needing to support the following on occasion:

### Example (Before)

```rb
class Invitation < ActiveRecord::Base
  scope :expired, -> { where.not(expired_at: nil).where(expired_at: ...Time.zone.now) }
  
  def expired?
    expired_at && expired_at <= Time.zone.now
  end
end

Invitate.create!(expired_at: 3.days.from_now)
```

This PR proposes a `strategy` option which allows an additional check against current time.

### Example (Proposal)

```rb
class Invitation < ActiveRecord::Base
  has_event :expire, strategy: :time_comparison
end

invitation = Invitate.create!(expired_at: 3.days.from_now)
invitation.expired? # => false

travel_to(4.days.from_now) do
  invitation.expired? # => true
end
```

Thoughts on adding this use case? Thanks!

pienkowb commented:
Hey @serogers, thanks for the contribution! 👍 I really like this idea. I'll review the code over the weekend.

pienkowb commented on lib/generators/active_record/event/event_generator.rb:
```diff
@@ -17,6 +17,8 @@ class EventGenerator < Rails::Generators::Base
       class_option :object, type: :string,
         desc: 'The name of the object'
+      class_option :strategy, type: :string,
+        desc: 'Define comparison strategy, presence or time_comparison'
```
Let's change the description to:

> The comparison strategy (presence or time_comparison)

pienkowb commented on spec/dummy/db/schema.rb:
```diff
@@ -10,10 +10,10 @@
 
-ActiveRecord::Schema.define(version: 2015_08_13_132804) do
-
+ActiveRecord::Schema[7.0].define(version: 2015_08_13_132804) do
```
Is `Schema[7.0]` supported by the older Rails versions? The gem should be compatible with ActiveRecord 4+.

pienkowb commented on spec/active_record/events_spec.rb:
```diff
@@ -71,4 +71,50 @@
     expect(task.reload).to be_completed
   end
+
+  context 'when strategy is time_comparison' do
```
`'when the strategy is time comparison'`

pienkowb commented on lib/active_record/events/method_factory.rb:
```diff
@@ -34,9 +34,13 @@ def class_methods
       attr_reader :naming
 
-      def define_predicate_method(module_, naming)
+      def define_predicate_method(module_, naming, strategy = :presence)
```
These methods are not called outside of this class, so we can remove default value from the arguments list.

pienkowb commented on README.md:
````diff
@@ -119,6 +119,28 @@ has_event :complete, field_name: :completion_time
+**Example:**
+
+```ruby
+task.completed_at = 1.hour.ago
+task.completed? # => true
+Task.completed  # => [...]
+
+task.completed_at = 1.hour.from_now
+task.completed? # => false
+Task.completed  # => []
````
For the sake of simplicity, I would remove the lines that demonstrate the use of scopes. Also, I think they won't work as expected unless the record is saved first.

pienkowb commented on spec/dummy/app/models/task.rb:
```diff
@@ -1,6 +1,7 @@
 class Task < ActiveRecord::Base
   has_event :complete
   has_events :archive, skip_scopes: true
+  has_events :expire, strategy: :time_comparison
```
Since we test `has_events` here (the plural variant), we can change ` has_events` to  `has_event` (singular) one line above.

pienkowb reviewed (commented):
The code looks really good! I've left some minor comments.

Let's also check Rubocop – it is complaining in CI: https://github.com/pienkowb/active_record-events/actions/runs/2387952260

serogers commented:
Thanks for the review @pienkowb! I've updated the PR addressing your feedback. One outstanding item is appeasing rubocop, with one item that might require some minor refactoring:

>  Metrics/AbcSize: Assignment Branch Condition size for define_inverse_scope_method is too high

I'll take a look into this and any other CI failures soon.

serogers commented:
CI is failing due to a rubocop violation in the generated `Task` model. Since this is generated code, rather than project code, perhaps it should be ignored? We can add `tmp/**/*` to the rubocop exclusion rules?

```
tmp/app/models/task.rb:2:81: C: Metrics/LineLength: Line is too long. [87/80]
  has_event :complete, field_type: :date, skip_scopes: true, strategy: :time_comparison
                                                                                ^^^^^^^
```

pienkowb commented:
> We can add `tmp/**/*` to the rubocop exclusion rules?

Let's do it 👍 

I have one more thought - how would the generator behave if we don't pass the `strategy` option? Ideally, it should skip the `strategy` parameter entirely (rather than adding `strategy: presence`).

serogers commented:
Okay I've updated the Rubocop rule to exclude that tmp directory. Testing the generator, it leaves out any options not provided, which seems ideal as you mentioned. 

While testing, I also uncovered an edge case in the generator:

```rb
# lib/generators/active_record/event/event_generator.rb:44
inject_into_file model_file_path, "\s\s#{macro}\n", after: /class.+\n/
```

This will insert the `has_event` line after any mention of `class`. The models I was testing against had several "class" strings, which caused duplicate `has_event` lines. See example below. I think we can fix it with a small change to the regex: `^class\s.+\n`. Would you like me to commit that change to this PR?

```rb
# frozen_string_literal: true

class User < ApplicationRecord # *** Adds has_event after this line
  belongs_to :owner, class_name: "Parent", optional: true # *** Adds has_event after this line

  def some_method(object)
    # class of should respond to enumerable # *** Adds has_event after this line
  end
end
```

pienkowb commented:
> Okay I've updated the Rubocop rule to exclude that tmp directory. Testing the generator, it leaves out any options not provided, which seems ideal as you mentioned.

Perfect 👍 

> This will insert the `has_event` line after any mention of `class`. The models I was testing against had several "class" strings, which caused duplicate `has_event` lines. See example below. I think we can fix it with a small change to the regex: `^class\s.+\n`. Would you like me to commit that change to this PR?

Good catch! Although I think the regex should be `^\s+class\s.+\n` to account for namespaces, for instance:

```ruby
module Admin
  class User < ActiveRecord::Base
    # ...
  end
end
```

Feel free to commit the change in this PR.

serogers commented:
Good catch! I've committed a change to the pattern and tested against a few files.

serogers commented:
Thanks @pienkowb!

pienkowb commented:
@serogers I've added you to the Contributors list: https://github.com/pienkowb/active_record-events/commit/c95dc165fee50eb320f3054ce9f690ce2a06e13b 🎉
